Amazon.QLDB.Driver/session/AsyncSessionPool.cs

```diff
@@ -0,0 +1,175 @@
+        /// <param name="retryHandler">Handling the retry logic of the execute call.</param>
+        /// <param name="maxConcurrentTransactions">The maximum number of sessions that can be created from the pool at any one time.</param>
+        /// <param name="logger">Logger to be used by this.</param>
+        public AsyncSessionPool(
+            Func<CancellationToken, Task<Session>> sessionCreator,
+            IAsyncRetryHandler retryHandler,
+            int maxConcurrentTransactions,
+            ILogger logger)
+        {
+            this.sessionPool = new BlockingCollection<AsyncQldbSession>(maxConcurrentTransactions);
+            this.poolPermits = new SemaphoreSlim(maxConcurrentTransactions, maxConcurrentTransactions);
+            this.sessionCreator = sessionCreator;
+            this.retryHandler = retryHandler;
+            this.logger = logger;
+        }
+
+        /// <summary>
+        /// Execute a function asynchronously in session pool.
+        /// </summary>
+        /// <typeparam name="T">The return type of the function.</typeparam>
+        ///
+        /// <param name="func">The function to be executed in the session pool. The operation can be cancelled.</param>
+        /// <param name="retryPolicy">The policy on retry.</param>
+        /// <param name="retryAction">The customer retry action. The operation can be cancelled.</param>
+        /// <param name="cancellationToken">
+        ///     A cancellation token that can be used by other objects or threads to receive notice of cancellation.
+        /// </param>
+        ///
+        /// <returns>The result from the function.</returns>
+        public async Task<T> Execute<T>(Func<AsyncTransactionExecutor, Task<T>> func, RetryPolicy retryPolicy, Func<int, Task> retryAction, CancellationToken cancellationToken = default)
+        {
+            AsyncQldbSession session = null;
+            try
+            {
+                session = await this.GetSession(cancellationToken);
+                return await this.retryHandler.RetriableExecute(
+                    ct => session.Execute(func, ct),
+                    retryPolicy,
+                    async ct => session = await this.StartNewSession(ct),
+                    async ct =>
+                    {
+                        this.poolPermits.Release();
+                        session = await this.GetSession(ct);
+                    },
+                    retryAction,
+                    cancellationToken);
+            }
+            finally
+            {
+                if (session != null)
+                {
+                    session.Release();
+                }
+            }
+        }
+
+        /// <summary>
+        /// Dispose the session pool and all sessions.
+        /// </summary>
+        public async ValueTask Dispose()
+        {
+            this.isClosed = true;
+            while (this.sessionPool.Count > 0)
+            {
+                await this.sessionPool.Take().Close();
```

Comment:
No need to await this, nor does it need to be async.

This is also something that can be in the shared base class.

Reply:
No longer async.

However since sessionPool is defined as `BlockingCollection<QldbSession>`/`BlockingCollection<AsyncQldbSession>`, the sessionPool stuff needs to be defined in both even though the code is the same. I could maybe get around this with generics but I'm not worrying about that right now.